### PR TITLE
core/trie: persist TrieJournal to journal file instead of kv database

### PR DESCRIPTION
**Pull request by jingjunLi**
### Description

This PR provides another way to persist TrieJournal by write it into a log file instead of a kv database.

Use `--journalFile` flag to enable/disable the journal file feature.

### Rationale

Persisting TrieJournal to KV database will cause problems below

- Additional WAL and compaction overhead in pebble/leveldb
- Big kv (>= 2GB) not supported by pebbledb. ([Size out of bounds when write kv >= 2GB](https://github.com/cockroachdb/pebble/issues/3333))

### Changes:

**Data Format**

Journal Version
Disk Root
Disk Layer
-- `**Layer Length**`
-- Root
-- State ID
-- Buffer Nodes
-- `**Layer Checksum**`
Diff Layer
-- `**Layer Length**`
-- Root
-- Block
-- Trie Nodes
-- Journal Accounts
-- `**Layer Checksum**`

Journal Path: geth/chaindata/ancient/state.journal
Checksum: SHA256

#### Function test
- Normal Scenario:
    - When starting up for the first time, if the journal does not exist, blocks are appended normally.
    - Upon receiving a **`kill -15`** signal, the journal is persisted with a size consistent with the file.
    - After the normal startup, blocks are appended normally.
    
image

    The loaded diff layer matches the persistence saved as described above.
    
- Abnormal Scenario: Manually truncate a portion of the snap file (**`geth/chaindata/ancient/state.journal`**) and restart.
    - When the number of blocks appended is less than 460,000 ("Rewound to block with state" number=0), the blockchain rewinds to block 0 and resumes appending blocks once completed.
    - When the number of blocks appended is greater than 460,000, the blockchain rewinds to a position slightly behind the head number, and resumes appending blocks once completed.
image

####  Performance test

Scenario: Using TestJournal, each layer has 4K and 8K account operations.
|            | WriteJournal | LoadJournal | Dirty State Size | 
|------------|--------------|-------------|------------------| 
| **New Journal WAL** | 6.149s | 12.496s | 1.19GiB | 
| Old KVDB        | 6.341s | 12.340s | 1.21GiB | 
| **New Journal WAL** | 11.724s | 20.525s | 1.85GiB | 
| Old KVDB        | 11.107s | 19.415s | 1.85GiB |  

Conclusion: The use of Journal WAL and storing trieJournalKey directly into kvdb results in nearly identical WAL write times during shutdown and WAL read/parsing times during startup.

### CODE REVIEW

File: triedb/pathdb/journal.go
Comment (Fynnss) [inline triedb/pathdb/journal.go:495]:
```
type Journal interface {
	// NewJournalWriter creates a new journal writer.
	NewJournalWriter() io.Writer

	// NewJournalReader creates a new journal reader.
	NewJournalReader() (*rlp.Stream, error)

	// JournalWriterSync flushes the journal writer.
	Write()

	// JournalDelete deletes the journal.
	Delete()

	// JournalClose closes the journal.
    Close()

	// JournalSize returns the size of the journal.
	Size() uint64
}
```

File: triedb/pathdb/journal.go
Comment (Fynnss) [inline triedb/pathdb/journal.go:556]: newJournalWriter

File: triedb/pathdb/journal.go
Comment (Fynnss) [inline triedb/pathdb/journal.go:566]: newJournalReader

File: cmd/utils/flags.go
Comment (Fynnss) [inline cmd/utils/flags.go:369]:
```suggestion
		Name:     "journal-file",
```

File: cmd/utils/flags.go
Comment (Fynnss) [inline cmd/utils/flags.go:370]: `Enable journal file to store the TrieJounal when shutdown`

File: eth/backend.go
Comment (Fynnss) [inline eth/backend.go:72]: trie.journal

**Comment by zzzckck** [comment]
The solution is ok for me, it would bring some advantages:
1.avoid the 2GB large KV limit of PebbleDB, as TrieJournal could > 2GB
2.avoid set large content into the PebbleDB, which would be deleted on restart, additional DBCompact cost. Log file is much more efficient.
I was a little bit concerned about the code intrusion, but I think the change is not too much and is acceptable.

It can be merged once there are >2 approval.

File: triedb/pathdb/journal.go
Comment (RenRick) [inline triedb/pathdb/journal.go:80]:
```suggestion
	journalBuf bytes.Buffer
```

File: triedb/pathdb/journal.go
Comment (RenRick) [inline triedb/pathdb/journal.go:81]:
```suggestion
	diskdb     ethdb.Database
```

File: triedb/pathdb/journal.go
Comment (RenRick) [inline triedb/pathdb/journal.go:84]:
```suggestion
// journalFile is used to store trie journal into a file.
```

**Review by RenRick — DISMISSED** [review]
LGTM

File: triedb/pathdb/journal.go
Comment (Fynnss) [inline triedb/pathdb/journal.go:77]: Split Journal Interface it into Writer and Reader

```
type JournalWriter interface {
	io.Writer

	Close()
	Size()
}

type JournalReader interface {
	io.Reader

	Close()
}
```

File: triedb/pathdb/journal.go
Comment (Fynnss) [inline triedb/pathdb/journal.go:90]:
```suggestion

```

**Review by buddh0 — APPROVED** [review]

**Review by Fynnss — APPROVED** [review]
LGTM

**Review by RenRick — APPROVED** [review]
LGTM

**Review by setunapo — APPROVED** [review]

**Review by zzzckck — APPROVED** [review]

**Pull request merged**
